### ISSUE TITLE
Spurious spaces in Czech

### ISSUE THREAD

**Issue by moewew** [reutenauer/polyglossia#423]
Via https://tex.stackexchange.com/q/547477/35864

Consider
```latex
\documentclass{article}

\usepackage{polyglossia}
\setmainlanguage{english}
\setotherlanguages{czech,spanish}

\begin{document}
A\textczech{B}C

A\textspanish{B}C
\end{document}
```
compiled with XeLaTeX the output is
![A(spaces)BC//ABC](https://user-images.githubusercontent.com/6755835/83494545-bb346e00-a4b6-11ea-8ef5-f58381b14491.png)

With LuaLaTeX the result is slightly better, but there is still a spurious space.

The obvious missing `%`s are
```diff
--- gloss-czech.ldf	2020-06-02 09:50:26.557694400 +0200
+++ gloss-czech-nospace.ldf	2020-06-02 09:49:40.262734100 +0200
@@ -77,7 +77,7 @@
       \chardef\czech@boundary=4095 %
       \def\newXeTeXintercharclass{%
         \e@alloc\XeTeXcharclass\chardef
-              \xe@alloc@intercharclass\m@ne\@ucharclass@boundary}
+              \xe@alloc@intercharclass\m@ne\@ucharclass@boundary}%
     \else
       \chardef\czech@boundary=255
     \fi
@@ -132,9 +132,9 @@
         \XeTeXcharclass `\a \czech@nonsyllabicpreposition
         \XeTeXcharclass `\I \czech@nonsyllabicpreposition
         \XeTeXcharclass `\i \czech@nonsyllabicpreposition
-        \XeTeXinterchartoks \czech@boundary \czech@nonsyllabicpreposition {\czech@interchartoks}
-        \XeTeXinterchartoks \czech@openpunctuation \czech@nonsyllabicpreposition {\czech@interchartoks}
-        \def\czech@nointerchartoks{\let\czech@interchartoks\czech@PreCSpreposition}
+        \XeTeXinterchartoks \czech@boundary \czech@nonsyllabicpreposition {\czech@interchartoks}%
+        \XeTeXinterchartoks \czech@openpunctuation \czech@nonsyllabicpreposition {\czech@interchartoks}%
+        \def\czech@nointerchartoks{\let\czech@interchartoks\czech@PreCSpreposition}%
         \def\czech@PreCSpreposition{\def\next{}%
            \ifnum\catcode`\ =10 % nothing will be done in verbatim
            \ifmmode % nothing in math
@@ -143,9 +143,9 @@
               \let\next\czech@ExamineCSpreposition
            \fi \fi
           \next%
-        }
-        \def\czech@ExamineCSpreposition ##1{##1\futurelet\next\czech@ProcessCSpreposition}
-        \def\czech@ProcessCSpreposition{\ifx\next\czech@XeTeXspace\nobreak\fi}
+        }%
+        \def\czech@ExamineCSpreposition ##1{##1\futurelet\next\czech@ProcessCSpreposition}%
+        \def\czech@ProcessCSpreposition{\ifx\next\czech@XeTeXspace\nobreak\fi}%
         \futurelet\czech@XeTeXspace{ }\czech@nointerchartoks
     \fi
 }
```

But that does not resolve everything. The remaining problematic line seems to be

https://github.com/reutenauer/polyglossia/blob/af00c264a4c694bb05e4c910912419281e88ea7d/tex/gloss-czech.ldf#L149

which I don't quite understand.

**Comment by jspitz** [comment]
Thanks, @moewew. I fixed the remaining issue by moving those definitions out of `\czech@vlna`.

`\futurelet\czech@XeTeXspace{ }\czech@nointerchartoks` always expands the space, and the `\czech@vlna` is called on language selection. Of course we only need to define those macros once, and it does not harm to define them unconditionally.

The issue was also in Slovak, which uses identical code for vlna.

**Comment by jspitz** [comment]
> and it does not harm to define them unconditionally.

On the other hand, we do not need then in luatex. Amended in d0f3da69c59.

**Comment by jspitz** [comment]
BTW as noted on stackexchange, there is also one spurious space with luatex. This is caused by the `luavlna` package. Reported at https://github.com/michal-h21/luavlna/issues/5

**Comment by jspitz** [comment]
> BTW as noted on stackexchange, there is also one spurious space with luatex. This is caused by the `luavlna` package

and has been fixed in `luavlna 0.1g 2020-06-02`:
https://www.ctan.org/ctan-ann/id/mailman.1720.1591194834.3866.ctan-ann@ctan.org
